batchupload/uploader.py

```diff
@@ -172,36 +184,173 @@ def up_all(in_path, cutoff=None, target='Uploaded', file_exts=None,
+        'general': common.LogFile(output_dir, '¤uploader.log')
+    }
+
+    # shortcut to the general/verbose logfile
+    flog = logs['general']
+
+    # @todo: filtering/removal of entries based on new input parameters
+
+    counter = 1
+    for url, data in info_datas.items():
+        if cutoff and counter > cutoff:
+            break
+
+        # verify that the file extension is ok
+        try:
+            ext = verify_url_file_extension(url, file_exts)
+        except common.MyError as e:
+            flog.write(e)
+            continue
+
+        # verify that info and output filenames are provided
+        if not data['info'] or not data['filename']:
+            flog.write(
+                '{url}: Found url missing one, or both, of info or output '
+                'filename'.format(
+                    url=url))
+            continue
+
+        # prepare upload
+        txt = make_info_page(data)
+        filename = '{filename}{ext}'.format(filename=data['filename'], ext=ext)
+
+        if test:
+            pywikibot.output(
+                'Test upload "{filename}" from {url} with the following '
+                'description: {txt}\n'.format(
+                    filename=filename, url=url, txt=txt))
+            continue
+        # stop here if testing
+
+        result = upload_single_file(
+            filename, url, txt, target_site, upload_if_badprefix=True)
+        if result.get('error'):
+            logs['error'].write(url)
+        elif result.get('warning'):
+            logs['warning'].write(url)
+        else:
+            logs['success'].write(url)
+        if verbose:
+            pywikibot.output(result.get('log'))
+
+        flog.write(result.get('log'))
+        counter += 1
+
+    for log in logs.values():
+        pywikibot.output(log.close_and_confirm())
+
+
+def verify_url_file_extension(url, file_exts, url_protocols=None):
+    """
+    Verify that an url contains a file extension and that it is allowed.
+
+    Also checks that the protocol is whitelisted.
+
+    @param url: the url to check
+    @param file_exts: tuple of allowed file extensions
+    @param url_protocols: tuple of allowed url protocols
+    @return: the file extension
+    @raises: common.MyError
+    """
+    url_protocols = url_protocols or URL_PROTOCOLS
+
+    protocol, _, rest = url.partition('://')
+    if protocol not in url_protocols:
+        raise common.MyError(
+            '{0}: Found url with a disallowed protocol'.format(url))
+
+    try:
+        ext = os.path.splitext(url)[1]
+    except IndexError:
+        raise common.MyError(
+            '{0}: Found url without a file extension'.format(url))
+    else:
+        if not ext:
+            raise common.MyError(
+                '{0}: Found url without a file extension'.format(url))
+
+    if ext not in file_exts:
+        raise common.MyError(
+            '{0}: Found url with a disallowed file extension'.format(url))
```

Comment:
Maybe add the extension in parentheses in the end of the message? It's repetition but could make the whole thing  easier to read, especially if the url is long and it's visually hard to immediately spot where it ends.